designs/2023-rule-options-defaults/README.md

```diff
@@ -0,0 +1,295 @@
+  - Drawback: `RuleContext` already doesn't factor in default options to the type system (https://github.com/typescript-eslint/typescript-eslint/issues/5439)
+
+Resolving the type system type resulting from a fixed JSON schema -with or without `default`- and a user-provided object can be complex.
+It is out of scope for this RFC to investigate heuristics for that problem.
+
+## Documentation
+
+We'll want to augment the following docs generators in:
+
+- The core ESLint docs: automating and therefore standardizing how rule docs pages phrase default values
+- typescript-eslint's docs: the same as the core ESLint docs
+- [`eslint-doc-generator`](https://github.com/bmish/eslint-doc-generator)
+
+We'll also want to mention this in the ESLint core custom rule documentation.
+
+## Drawbacks
+
+- This increases both conceptual and implementation complexity around rule options
+- As a future breaking change, this may break user rules - especially private rules we don't have visibility to
+
+## Backwards Compatibility Analysis
+
+The new `useRecursiveOptionDefaults` option being opt-in means this would not be a breaking change in the current version of ESLint.
+This RFC's intent is for that option to gradually be removed over time:
+
+1. ESLint v8: opt-in (default to `false`)
+2. ESLint v9: opt-out (default to `true`)
+3. ESLint v10: remove both `useRecursiveOptionDefaults` and `context.optionsRaw`
+
+Changing `useRecursiveOptionDefaults` to opt-out would be a breaking change as it impacts what values are passed to the `create` method of rules.
+That breaking change will impact rules which rely on the exact length or existing of properties in `context.options`.
+Rules that wish to preserve the legacy options behavior can switch to referring to the new `context.optionsRaw` property.
+
+In the draft PR containing this change, only 3 test files -rule tests for `indent`, `max-len`, and `no-multiple-empty-lines`- failed during `npm run test`.
+All were fixed with a direct find-and-replace from `context.options` to `context.optionsRaw` in those rules' implementations.
+
+## Alternatives
+
+### Non-Recursive Object Defaults
+
+Ajv provides its own support for `default` values with a [`useDefaults` option](https://ajv.js.org/guide/modifying-data.html#assigning-defaults).
+It's used in ESLint core already, in [`lib/shared/ajv.js`](https://github.com/eslint/eslint/blob/22a558228ff98f478fa308c9ecde361acc4caf20/lib/shared/ajv.js#L21).
+Sticking Ajv's own defaults logic would bring two benefits:
+
+- Aligning ESLint schema defaulting with the standard logic used by Ajv
+- Avoiding a custom defaulting implementation in ESLint by using Ajv's instead
```

Comment:
I'd prefer this alternative. Schemas can be complex, seems better to let Ajv do the work it was made for. We'd just probably want to fix how we handle top-level arrays where defaults currently don't apply, which is I believe an unintentional omission.

Reply:
Downside of this: we'd be locked into Ajv (even more than we are already).